controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -399,45 +420,56 @@ public String getScopeName() {
+    public CompletableFuture<Pair<TxnStatus, Integer>> sealTransaction(final UUID txId, final boolean commit,
                                                         final Optional<Integer> version) {
-        return verifyLegalState(checkTransactionStatus(txId)
-                .thenCompose(x -> {
-                    if (commit) {
-                        switch (x) {
-                            case OPEN:
-                                return sealActiveTx(txId, true, version).thenApply(y -> TxnStatus.COMMITTING);
-                            case COMMITTING:
-                            case COMMITTED:
-                                return CompletableFuture.completedFuture(x);
-                            case ABORTING:
-                            case ABORTED:
-                                throw new OperationOnTxNotAllowedException(txId.toString(), "seal");
-                            default:
-                                throw new TransactionNotFoundException(txId.toString());
-                        }
-                    } else {
-                        switch (x) {
-                            case OPEN:
-                                return sealActiveTx(txId, false, version).thenApply(y -> TxnStatus.ABORTING);
-                            case ABORTING:
-                            case ABORTED:
-                                return CompletableFuture.completedFuture(x);
-                            case COMMITTING:
-                            case COMMITTED:
-                                throw new OperationOnTxNotAllowedException(txId.toString(), "seal");
-                            default:
-                                throw new TransactionNotFoundException(txId.toString());
-                        }
-                    }
-                }));
+        CompletableFuture<Pair<TxnStatus, Integer>> future = verifyLegalState(getTransactionEpoch(txId)
+                .thenCompose(epoch -> sealActiveTxn(epoch, txId, commit, version)))
+                .exceptionally(ex -> new ImmutablePair<>(handleDataNotFoundException(ex), null));
+        return future.thenCompose(pair -> pair.getLeft() == TxnStatus.UNKNOWN ?
+                validateCompletedTxn(txId, commit, "seal").thenApply(status -> new ImmutablePair<>(status, null)) :
+                CompletableFuture.completedFuture(pair));
+    }
+
+    public CompletableFuture<Pair<TxnStatus, Integer>> sealActiveTxn(final int epoch, final UUID txId,
+                                                                     final boolean commit,
+                                                                     final Optional<Integer> version) {
+        return getActiveTx(epoch, txId).thenCompose(data -> {
+            ActiveTxnRecord txnRecord = ActiveTxnRecord.parse(data.getData());
+            int dataVersion = version.isPresent() ? version.get() : data.getVersion();
+            TxnStatus status = txnRecord.getTxnStatus();
+            if (commit) {
```

Comment:
Instead of two switch/case blocks, would it be better to have one switch/case block and have `if (commit) ` for the cases that require it, `COMMITTING/COMMITTED` and `ABORTING/ABORTED`?